### DISCUSSION TITLE
Autostart python program

### DISCUSSION THREAD

**Discussion by pgraykowski69** [micropython/micropython#9326] [Using MicroPython]
Hi. I'm trying to fully understand the boot process for MicroPython. I'm on the RP2 port.

By default there is no 'boot.py' nor 'main.py' on my board.  I moved my program to /main.py and it runs at powerup (great!)

In the source:

port/rp2/main.c 

It loads Micropython and then does a few things...

if file _boot_fat.py or _boot_.py exists these are executed (file system). Both exist in the git repo so the defines that gate them determine which one. While I haven't investigated far enough to determine the differences, I get the the gist that a file system is loaded (makes sense to be able to look for boot.py and main.py).

next it looks for boot.py and executes
next  it looks for main.py and executes
It appears these files must be in the root of the filesystem

I have found some brief mentions of boot.py and main.py in various searches but no real documentation of the suggested flow. I renamed my program main.py and it does run at boot. I'd really like to understand the process here. i.e. what should I put in boot.py versus main.py.   Hopefully I missed the documentation but this is critical to know the bringup sequence for running the board w/o a host computer.  

Thanks in advance.

**Comment by robert-hh** [comment]
Either `_boot.py` or  `_boot_fat.py` are needed to create a file system. For `boot.py` it's LFS, for `_boot_fat.py` it's FAT. The setting for MSC support rules, which of both is used.
`boot.py` and `main.py` are executed one after the other, if they exists. There is not mandatory rule, what put where. The intention is to put all board set-up code into `boot.py`, and the application or start of application into `main.py`. Some ports have the REPL resp USB interface not active when `boot.py` is executed. So any code the requires interaction over USB must be placed into main.py.
It is also good practice not to put all of your code in main.py, but just start it from there, e.g. with `import mycode`.

**Comment by pgraykowski69** [comment]
Thanks.  Related question, I'm using a prebuilt Micropython install. Is there a runtime way to check which filesystem is installed? The 2040 Datasheet mentions littlefs but the code clearly has the option to load either Vfslfs2 or VfsFat.  I have no idea what defines were used in the building of the release though from what I've read littlefs seems to make the most sense for embedded flash. If an external SDCard then FAT. I"m assuming the Micropython os and fileio related commands layer on top of the underlying file system - i.e. no special libraries are needed.

**Comment by robert-hh** [comment]
There is a little Python script below made by @dpgeorge which shows the type of file system and it's properties. The information is from the boot sector.  The standard build of the `RP2040` uses LFS. As far as I know, only the `Arduino Nano Connect` has MSC support enabled by default and uses therefore FAT:
```

"""
Dump info about a filesystem.
"""

import sys, struct

def pr(addr, label, data):
    if type(data) == bytes or type(data) == bytearray:
        data = ':'.join('%02x' % b for b in data)
    print('%04x  %-32s %s' % (addr, label, data))

def decode_bootsec_fat(b):
    print('FAT filesystem')
    pr(0, 'jump', b[0:3])
    pr(3, 'OEM name', str(b[3:11], 'ascii'))
    pr(11, 'sector size (bytes)', struct.unpack_from('<H', b, 11)[0])
    pr(13, 'cluster size (sectors)', struct.unpack_from('<B', b, 13)[0])
    pr(14, 'reserved area (sectors)', struct.unpack_from('<H', b, 14)[0])
    pr(16, 'number of FATs', struct.unpack_from('<B', b, 16)[0])
    pr(17, 'size of root dir area', struct.unpack_from('<H', b, 17)[0])
    pr(19, 'volume size (sectors)', struct.unpack_from('<H', b, 19)[0])
    pr(21, 'media descsriptor', hex(struct.unpack_from('<B', b, 21)[0]))
    pr(22, 'FAT size (sectors)', struct.unpack_from('<H', b, 22)[0])
    pr(24, 'track size (sectors)', struct.unpack_from('<H', b, 24)[0])
    pr(26, 'number of heads', struct.unpack_from('<H', b, 26)[0])
    pr(28, 'volume offset (sectors)', struct.unpack_from('<L', b, 28)[0])
    pr(32, 'volume size (32-bit) (sectors)', struct.unpack_from('<I', b, 32)[0])
    pr(36, 'physical drive number', struct.unpack_from('<B', b, 36)[0])
    pr(37, 'error flag', hex(struct.unpack_from('<B', b, 37)[0]))
    pr(38, 'extended boot signature', hex(struct.unpack_from('<B', b, 38)[0]))
    pr(39, 'volume serial number', hex(struct.unpack_from('<L', b, 39)[0]))
    pr(43, 'volume label', str(b[43:51], 'ascii'))
    pr(54, 'filesystem type', str(b[54:62], 'ascii'))
    pr(510, 'signature', hex(struct.unpack_from('<H', b, 510)[0]))


def decode_bootsec_lfs1(b):
    print("Littlefs v1 filesystem")
    pr(24, "type", struct.unpack_from("<I", b, 24)[0])
    pr(25, "elen", struct.unpack_from("<I", b, 25)[0])
    pr(26, "alen", struct.unpack_from("<I", b, 26)[0])
    pr(27, "nlen", struct.unpack_from("<I", b, 27)[0])
    pr(28, "block_size", struct.unpack_from("<I", b, 28)[0])
    pr(32, "block_count", struct.unpack_from("<I", b, 32)[0])
    pr(36, "version", hex(struct.unpack_from("<I", b, 36)[0]))
    pr(40, "magic", str(b[40:40 + 8], 'ascii'))


def decode_bootsec_lfs2(b):
    print("Littlefs v2 filesystem")
    pr(8, "magic", str(b[8:8 + 8], "ascii"))
    pr(20, "version", hex(struct.unpack_from("<I", b, 20)[0]))
    pr(24, "block_size", struct.unpack_from("<I", b, 24)[0])
    pr(28, "block_count", struct.unpack_from("<I", b, 28)[0])
    pr(32, "name_max", struct.unpack_from("<I", b, 32)[0])
    pr(36, "file_max", struct.unpack_from("<I", b, 36)[0])
    pr(40, "attr_max", struct.unpack_from("<I", b, 40)[0])


def decode_bootsec(b):
    if b[40:48] == b"littlefs":
        decode_bootsec_lfs1(b)
    elif b[8:16] == b"littlefs":
        decode_bootsec_lfs2(b)
    else:
        decode_bootsec_fat(b)

def main():
    if sys.platform == 'pyboard':
        import pyb
        #sd = pyb.SDCard()
        sd = pyb.Flash(start=0)
        bootsec = bytearray(512)
        sd.readblocks(0, bootsec)
    elif sys.platform in ('esp8266', 'esp32'):
        import esp
        bootsec = bytearray(512)
        esp.flash_read(esp.flash_user_start(), bootsec)
    else:
        with open(sys.argv[1], 'rb') as f:
            bootsec = f.read(512)
    decode_bootsec(bootsec)

main()
```
